### PR TITLE
[FLINK-23272][web] Add custom netty HTTP request inbound/outbound handlers

### PR DESCRIPTION
**Pull request by gaborgsomogyi**
## What is the purpose of the change

At the moment there is no possibility to influence netty HTTP inbound/outbound requests. This could be useful from many perspectives. A good example (amongst others) is authentication/authorization. Here are the exact examples:
* With an outbound custom netty handler one can modify the HTTP request by adding authentication information.
* With an inbound custom netty handler one can do authentication/authorization.

In this PR I've added 2 generic APIs in order to handle inbound and outbound HTTP requests.
Please see [FLIP-181](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=184616200) for further details.

## Brief change log

* Added. `InboundChannelHandlerFactory`
* Added. `OutboundChannelHandlerFactory`
* `TestRestServerEndpoint` and `RestClient` are now getting Flink `Configuration` instances.

## Verifying this change

* Existing automated tests
* There are 2 not yet published external repositories with automated tests using the newly added APIs:
  * Basic authentication
  * Kerberos authentication

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes, introduced 2 `@Experimental` APIs
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs + API usage examples (intended to be published when APIs are in place)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/InboundChannelHandlerFactory.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    /**
+     * Creates new instance of {@link ChannelHandler}
+     *
+     * @param configuration The Flink {@link Configuration}.
+     * @param responseHeaders The response headers.
+     * @return {@link ChannelHandler} or null if no custom handler need to be created.
+     * @throws ConfigurationException Thrown, if the handler configuration is incorrect.
+     */
+    ChannelHandler createHandler(Configuration configuration, Map<String, String> responseHeaders)
```
Comment (gyfora) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/InboundChannelHandlerFactory.java:53]: I agree with @tillrohrmann, this should return Optional<ChannelHandler> to make the behaviour explicit

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/OutboundChannelHandlerFactory.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    int priority();
+
+    /**
+     * Creates new instance of {@link ChannelHandler}
+     *
+     * @param configuration The Flink {@link Configuration}.
+     * @return {@link ChannelHandler} or null if no custom handler needs to be created.
+     * @throws ConfigurationException Thrown, if the handler configuration is incorrect.
+     */
+    ChannelHandler createHandler(Configuration configuration) throws ConfigurationException;
```
Comment (gyfora) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/OutboundChannelHandlerFactory.java:51]: I agree with @tillrohrmann, this should return Optional<ChannelHandler> to make the behaviour explicit

**Review by gyfora — CHANGES_REQUESTED** [review]
Added some minor comments, also please add a test that validates that handlers are actually added and in the correct order according to their priority.

We can use some test/mock handlers that we can easily validate :)

**Comment by gaborgsomogyi** [comment]
As a start resolved the conflict.

**Comment by gaborgsomogyi** [comment]
The tests are not yet testing the netty chain so discovering how is it possible to do that w/o horror complex code...

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/InboundChannelHandlerFactory.java (hunk 1 above)
Comment (gaborgsomogyi) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/InboundChannelHandlerFactory.java]: Changed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/OutboundChannelHandlerFactory.java (hunk 2 above)
Comment (gaborgsomogyi) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/OutboundChannelHandlerFactory.java]: Changed.

**Comment by gaborgsomogyi** [comment]
@gyfora added proper tests which are testing the loaded factories behavior but this has increased the PR size significantly. I don't think I can come up w/ more simple tests. If somebody has any idea please share.

**Comment by gyfora** [comment]
@flinkbot run azure

**Comment by gyfora** [comment]
@flinkbot approve all

**Comment by gaborgsomogyi** [comment]
`SQLClientSchemaRegistryITCase` failed but it's passing in my local environment so tent to think it's flaky.

**Comment by gaborgsomogyi** [comment]
@flinkbot run azure

**Comment by gaborgsomogyi** [comment]
`BlockingShuffleITCase` failed which is another flaky test. Filing jira-s to highlight the flakyness of these tests.

**Comment by gaborgsomogyi** [comment]
Filed FLINK-23665 and FLINK-23666.

**Comment by tillrohrmann** [comment]
@gyfora it would be nice to update the JIRA tickets when merging a PR. Otherwise it is hard to find the paper trail when looking things up. Concretely, this PR seems to reference FLINK-21108 as in the commit but the PR title states that it is FLINK-23272. Moreover, FLINK-21108 has been closed as a duplicate more than a month ago.

**Comment by gyfora** [comment]
Yea, we updated the ticket but unfortunately wrong ticket was referenced in the commit. We were so preoccupied with the CI problems that we did not notice this. :(

**Comment by tillrohrmann** [comment]
Np. I think it should be fine if you update the respective Jira tickets accordingly. Thanks!

**Pull request closed without merging**